File: solutions/clojure/run-length-encoding/1/src/run_length_encoding.clj
(ns run-length-encoding)

(defn encode [encoded token]
  (let [current (encoded :current)
        result (encoded :result)
        count (encoded :count)
        token (str token)]
    (cond
      (= current token) (assoc encoded :count (inc count))
      :else (assoc encoded
                   :result (cond
                             (empty? current) ""
                             :else (str result (if (= count 1) "" count) current))
                   :count 1
                   :current token))))

(defn run-length-encode
  "encodes a string with run-length-encoding"
  [plain-text]
  (let [encoded
        (reduce encode {:result "" :current "" :count 0} plain-text)
        final-count (encoded :count)]
    (str (encoded :result) (if (< final-count 2) "" final-count) (encoded :current))))

(defn decode [decoded token]
  (let [result (decoded :result)
        number (decoded :number)
        token (str token)]
    (cond
      (re-seq #"[0-9]" token) (assoc decoded :number (str number token))
      :else (assoc decoded :result (cond
                                     (= number "") (str result token)
                                     :else (str result (apply str (repeat (Integer. number) token))))
                           :number ""))))

(defn run-length-decode
  "decodes a run-length-encoded string"
  [cipher-text]
  (let [decoded (reduce decode {:result "" :number ""} cipher-text)]
    (decoded :result)))

File: solutions/julia/run-length-encoding/1/run-length-encoding.jl
function encode(s)
    if s == ""
        return ""
    end
    encoded = ""
    seen = ""
    seen_count = 0
    for (fst, snd) in zip(s, string(s[2:end], "."))
        if seen == fst
            seen_count += 1
        else
            seen = fst
            seen_count = 1
        end
        if fst != snd
            if seen_count == 1
                encoded = string(encoded, seen)
            else
                encoded = string(encoded, seen_count, seen)
            end
        end
    end
    return encoded
end

function decode(s)
    if s == ""
        return ""
    end
    decoded = ""
    count = "0"
    for (fst, snd) in zip(s, string(s[2:end], "."))
        if snd == "."
            decoded = string(decoded, fst)
        elseif isdigit(fst)
            count = string(count, fst)
        elseif ~isdigit(fst)
            if count == "0"
                count = "1"
            end
            decoded = string(decoded, repeat(fst, parse(Int, count)))
            count = "0"
        end
    end
    return decoded
end
